Add TP for QUARKUS-5613 OIDC Response filter mechanism (#193)
quarkus-qe/quarkus-test-plans: QUARKUS-5613.md

## Links to the resources
- JIRA ticket: https://issues.redhat.com/browse/QUARKUS-5613
- GH PRs related to the JIRA issue: https://github.com/quarkusio/quarkus/pull/43283 https://github.com/quarkusio/quarkus/issues/42117
- TP/TD PR: https://github.com/quarkus-qe/quarkus-test-plans/pull/193
- Guides modified and related:
  * https://quarkus.io/guides/security-openid-connect-client-reference
  * https://quarkus.io/guides/security-oidc-code-flow-authentication
  * https://quarkus.io/guides/security-openid-connect-client-registration

## Scope of the testing

Our main testing scope is to ensure that the new OIDC response filter mechanism works as intended for various OIDC endpoints, ensuring that is possible intercept and handle responses
from various OIDC endpoints. This complements the already available OidcRequestFilter functionality.
This kind of validation is possible by asserting that:
    - response body can be read and serialized,
    - status code is correct,

>     - response headers are correct

Reviewer: first 2 list items are ended with a comma and last 2 items are not. I have no idea what is common in English world, but I'd do it stuff in uniform pattern. so let's add command here as well and behind the last one dot, or no commas, thanks

    - and request properties matches the ones recorded in the request filter

This scope will be covered on areas related to:
   - Security - ensuring integration with OIDC provider (as Keycloak) plus Quarkus OIDC and OIDC Client extensions.
   - Logging /Tracing -  to enable logs only at finer levels for our specific filter classes to confirm that the filter is triggered. This helps debug the response filter’s internal behavior.

The test coverage will be verified for JVM, native mode, and Openshift.
The testing include OIDC Client (quarkus-oidc), OIDC Client (quarkus-oidc-client) and OIDC client registration (quarkus-oidc-client-registration) extensions.

### What to Test - Scenarios

> We are focusing on these three scenarios (Token,User info, Client Registration) as they represent the most common endpoints for Quarkus OIDC users.

Reviewer: ```suggestion
We are focusing on these three scenarios (Token, UserInfo, Client Registration) as they represent the most common endpoints for Quarkus OIDC users.
```

> most common endpoints

I am really no OIDC expert, so take it with reservations, but: I doubt that the client registration is most common. I think it is rather infrequent scenario. You are ignoring scenarios that are truly most common as JWKS calls ...

Author: Include JWKS scenario and dropping the client-registration scenario.

#### Token Endpoint Response Filtering
 - Verify that a custom OIDC response filter (TokenEndpointResponseFilter) is triggered immediately after a successful token refresh.
 - In particular this scenario will use `quarkus-oidc` and `quarkus-oidc-client`

>  - Validation

Reviewer: ```suggestion
 - Refresh Request Validation
```

you are very specific, it is fine, but what you describe is in Quarkus OIDC only valid for one of calls, one of application type when access token expires and code flow is used while token refresh is supported.

Reviewer: basically what I am trying to say: title is `Token Endpoint Response Filtering` but you are verifying one of situations when token endpoint is called. That is fine, but I think you need to be clear that it is a situation that you decided to test.

    * Filter runs and logs claim details(`sub`, `preferred_username`)
    * JSON body response includes the details.
    * Logs show filter intercepting `/userinfo` response
Note:  We do not test Keycloak’s user info correctness; only that the filter sees the response code/body.

#### Client Registration Endpoint Response Filtering
- Create a filter(OidcEndpoint.Type.CLIENT_REGISTRATION) intercepts responses to client registration calls.
- Validation
  * Registration data, for instance check filter logs newly registered `client_id` or `client_name`
  * Check if filter can detect a missing invalid property in the returned JSON
  * Error handling, if Keycloak return a 401 or 4xx the filter logs or throws an exception appropriately

## Getting familiar with the feature

- Check and looking through the guides:

 * https://quarkus.io/guides/security-openid-connect-client-reference

 * https://quarkus.io/guides/security-oidc-code-flow-authentication

 * https://quarkus.io/guides/security-openid-connect-client-registration

## Community status of the extension/feature:

> All the extensions used in the module are stable and supported.

Reviewer: `All the extensions used in the module are stable and supported.` is not true anymore as it is experimental but in product and in community.

## Links to the resources

> - JIRA issue: https://issues.redhat.com/browse/QUARKUS-5613

Reviewer: ```suggestion
- JIRA ticket: https://issues.redhat.com/browse/QUARKUS-5613
```

> - GH PRs related to the Jira issue: https://github.com/quarkusio/quarkus/pull/43283 https://github.com/quarkusio/quarkus/issues/42117 

Reviewer: ```suggestion
- GH PRs related to the Jira issue: https://github.com/quarkusio/quarkus/pull/43283, https://github.com/quarkusio/quarkus/issues/42117 
```

- Guides modified and related:
  https://quarkus.io/guides/security-openid-connect-client-reference
  https://quarkus.io/guides/security-oidc-code-flow-authentication
  https://quarkus.io/guides/security-openid-connect-client-registration

> - TP/TD PRs:

Reviewer: ```suggestion

```
not necessary, if anyone needs to track it down, all you need to do is to link TD to the Quarkus JIRA ticket.

Reviewer: also, linking TP PR :-)

## Scope of the testing

Our main testing scope is to ensure that the new response filter mechanism works as intended for various OIDC endpoints, ensuring that is possible intercept and handle responses
from various OIDC endpoints.

This complements the already available OidcRequestFilter functionality.

> Integration tests will be on the next module in our Test suite:    

Reviewer: that belongs to section `## Impact on test suites and testing automation`

Reviewer: or to the `Automated test development` section that you already have.

> Integration tests will be on the next module in our Test suite:    

Reviewer: What do you mean by the `on the next module`? You named explicitly which module, are they going to be in same module, different module, or module sorted next to the one where request filters are?

I suspect it is just a typo.

 security/keycloak-oidc-client-reactive-extended (https://github.com/quarkus-qe/quarkus-test-suite/tree/main/security/keycloak-oidc-client-reactive-extended).
And the scope will be covered on areas related to:
   - Security

>    - Logging /Tracing - Enabling TRACE level logs in properties will capture relevant info for our tests

Author: I was playing and creating a test and it's only for specific filter classes that ensure that the filter is triggered as expected and the data is being processed by filters...That's why I put it there.

>    - Logging /Tracing - Enabling TRACE level logs in properties will capture relevant info for our tests

Reviewer: I can't see relation to OIDC response filter, but it is absolutely fine as long as you change this text to mention that tracing will only be enabled for a specific class (the response filter class) to test something that customers could do

The test coverage will be verified for JVM, native mode, and Openshift.

> The testing include both OIDC Client (quarkus-oidc) and OIDC Client (quarkus-oidc-client) extensions.

Reviewer: You plan to test `Client Registration Endpoint Response Filtering` however that is a different extension: `oidc-client-registration`.

> ### What to Test - Scenarios

Reviewer: In other words, you will test exactly what you see already tested in Quarkus test coverage, but there are other scenarios where response filter is applied like `TOKEN_REVOCATION`, `JWKS`, `INTROSPECTION`, ...

I can respect your choice to test exactly these 3, but I think it can't be just stated, it must be reasoned WHY you chose these and I don't see it in the TP. Please kindly say in the TP something like _I chose these 3 because..._

#### Token Endpoint Response Filtering

>  - Verify that a custom OIDC response filter (TokenEndpointResponseFilter) is triggered immediately after a successful token refresh from Keycloak.

Reviewer: btw I do see at line 25 following sentence `The testing include both OIDC Client (quarkus-oidc) and OIDC Client (quarkus-oidc-client) extensions.` it is just that you will probably not test it for all 3 response types as for example client registration will not be fired in the OIDC or OIDC Client.

 - Validation
    * Ensure that filter logs messages contains "token have been refreshed".
    * Verify 'OidcResponseContext' exposes the correct property grant_type=refresh_token

>     * Assert that the response includes "refresh_token"

Reviewer: this `Validation` section proposes to verify what Keycloak sends, however that is Keycloak problem, not ours. We need to verify that validation of OIDC provider response is possible for body, headers, that request context fits the request filter and that logging is possible (as that is only documented thing...). So you can still test these scenarios to achieve that, but please rephrase whole `Validation` point. Thanks

> ####  User Info Endpoint Response Filtering

Reviewer: ```suggestion
####  UserInfo Endpoint Response Filtering
```
https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

Reviewer: I am not 100 % sure, but that is what specs does, you can ignore that, your choice

> - Demonstrate that an OIDC response filter can intercept the User Info endpoint response, for instance verifying user claims.

Reviewer: ```suggestion
- Demonstrate that an OIDC response filter can intercept the User Info endpoint response, for instance verifying token claims.
```

> - Demonstrate that an OIDC response filter can intercept the User Info endpoint response, for instance verifying user claims.

Reviewer: ```suggestion
- Demonstrate that an OIDC response filter can intercept the User Info endpoint response, for instance verifying UserInfo claims.
```
https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

-  Validation
    * Filter runs and logs claim details(sub,preferred_username)
    * JSON body response includes the details.

>     * Logs show filter intercepting '/userinfo' response

Reviewer: You plan to focus on Keycloak, so the path will be something like `/realms/{realm-name}/protocol/openid-connect/userinfo`, it is fine to mention just `/userinfo`, but I'd probably mention something about subpath or like that....

#### Client Registration Endpoint Response Filtering
- Create a filter(OidcEndpoint.Type.CLIENT_REGISTRATION) intercepts responses to client registration calls.
- Validation
  * Registration data, for instance check filter logs newly registered clint_id or client_name
  * Check if filter can detect a missing invalid property in the returned JSON
  * Error handling, if Keycloak return a 401 the filter log propagate it correctly.

> All scenarios will be tested in both JVM and Native mode on baremetal and Openshift.

Reviewer: you already mentioned this on line 23, once is enough

## Getting familiar with the feature

> - Focus on exploratory testing of the Keycloak and OIDC testing modules

Reviewer: That sounds very wrong. You fight for users, you are user champion! Do you expect Quarkus users to learn how to use this feature by reading Quarkus tests? Please focus on customer perspective.

Reviewer: e.g. documentation

Author: You're right, I am removing that sentence that was for me exploring what we have about it...

> - Check and looking thorough the guides:

Reviewer: ```suggestion
- Check and looking through the guides:
```

just fixing a typo, but the sentence is probably not good either. I am not proposing different wording as I am terrible at this.

>   https://quarkus.io/guides/security-openid-connect-client-reference

Reviewer: ```suggestion
*  https://quarkus.io/guides/security-openid-connect-client-reference
```

Reviewer: same goes for other items https://www.markdownguide.org/basic-syntax/#unordered-lists

  https://quarkus.io/guides/security-oidc-code-flow-authentication

  https://quarkus.io/guides/security-openid-connect-client-registration

## Community status of the extension/feature:

> All the extensions used in the module are stable and supported.

Reviewer: Oidc client extension is experimental upstream https://code.quarkus.io/?e=oidc-client-registration&extension-search=oidc-client-regi so either change scope of testing or this section, thanks

Author: Yeah Oidc-client-registration extension is experimental I will update it on the TP, thanks

Reviewer 2: Just a reminder, that `Experimental community status should be a showstopper for full or technology preview product support.` Is this extension required for filtering functionality? If yes, we should push for it to be stabilised (or drop the planned support). If no, then why do we need them?

Reviewer: It is not required, it is just a convenient way to register OIDC clients dynamically. It doesn't have direct relation to this new response filter, it is just one of events and @jcarranzan is  not testing all the events anyway.

 - quarkus-oidc (supported)
 - quarkus-oidc-client(supported)
 - quarkus-rest-client-oidc-filter (supported)

>  - quarkus-resteasy-client-oidc-filter (supported)

Reviewer: If you are only going to test one REST impl., I'd go for Quarkus REST client because that is recommended stack.

Author: Ok, removing quarkus-resteasy-client-oidc-filter extension, thanks

## Automated test development
Extend this test coverage existing security/keycloak-oidc-client-reactive-extended module:
- Implement tests that cover all relevant endpoints: token/refresh, /userinfo, /client-registration.

> - This validation is possible by asserting that:

Reviewer: this looks good, but it belongs to the scope of testing
